pyop2/mpi.py

```diff
@@ -118,63 +135,272 @@ def delcomm_outer(comm, keyval, icomm):
+
+    :returns: An mpi4py communicator."""
+    assert not is_pyop2_comm(comm_in)
+
+    # Check if communicator has an embedded PyOP2 comm.
+    comm_out = comm_in.Get_attr(innercomm_keyval)
+    if comm_out is None:
+        # Haven't seen this comm before, duplicate it.
+        comm_out = comm_in.Dup()
+        comm_in.Set_attr(innercomm_keyval, comm_out)
+        comm_out.Set_attr(outercomm_keyval, comm_in)
+        # Name
+        # TODO: replace id() with .py2f() ???
+        comm_out.Set_name(f"{comm_in.name or id(comm_in)}_DUP")
+        # Refcount
+        comm_out.Set_attr(refcount_keyval, [0])
+        incref(comm_out)
+        # Remember we need to destroy it.
+        dupped_comms.append(comm_out)
+    elif is_pyop2_comm(comm_out):
+        # Inner comm is a PyOP2 comm, return it
+        incref(comm_out)
+    else:
+        raise ValueError("Inner comm is not a PyOP2 comm")
     return comm_out
 
 
-# Comm used for compilation, stashed on the internal communicator
-compilationcomm_keyval = MPI.Comm.Create_keyval()
+@collective
+def create_split_comm(comm):
+    if MPI.VERSION >= 3:
+        debug("Creating compilation communicator using MPI_Split_type")
+        split_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)
+        debug("Finished creating compilation communicator using MPI_Split_type")
+    else:
+        debug("Creating compilation communicator using MPI_Split + filesystem")
+        import tempfile
+        if comm.rank == 0:
+            if not os.path.exists(configuration["cache_dir"]):
+                os.makedirs(configuration["cache_dir"], exist_ok=True)
+            tmpname = tempfile.mkdtemp(prefix="rank-determination-",
+                                       dir=configuration["cache_dir"])
+        else:
+            tmpname = None
+        tmpname = comm.bcast(tmpname, root=0)
+        if tmpname is None:
+            raise CompilationError("Cannot determine sharedness of filesystem")
+        # Touch file
+        debug("Made tmpdir %s" % tmpname)
+        with open(os.path.join(tmpname, str(comm.rank)), "wb"):
+            pass
+        comm.barrier()
+        import glob
+        ranks = sorted(int(os.path.basename(name))
+                       for name in glob.glob("%s/[0-9]*" % tmpname))
+        debug("Creating compilation communicator using filesystem colors")
+        split_comm = comm.Split(color=min(ranks), key=comm.rank)
+        debug("Finished creating compilation communicator using filesystem colors")
+    # Name
+    split_comm.Set_name(f"{comm.name or id(comm)}_COMPILATION")
+    # Refcount
+    split_comm.Set_attr(refcount_keyval, [0])
+    incref(split_comm)
+    return split_comm
 
 
 def get_compilation_comm(comm):
     return comm.Get_attr(compilationcomm_keyval)
 
 
 def set_compilation_comm(comm, inner):
-    comm.Set_attr(compilationcomm_keyval, inner)
+    """Set the compilation communicator.
+
+    :arg comm: A PyOP2 Communicator
+    :arg inner: The compilation communicator
+    """
```

Comment:
This docstring doesn't make it clear why you should be passing in two communicators.

Reply:
fixed